dev/Pager/PrototypePager/PrototypePager.Events.cs

```diff
@@ -0,0 +1,185 @@
+            }
+        }
+
+        private static void OnPreviousPageButtonVisibilityChanged(DependencyObject sender, DependencyPropertyChangedEventArgs args)
+        {
+            ButtonVisibilityMode prevPageButtonVisibilityMode = (ButtonVisibilityMode)args.NewValue;
+            PrototypePager pager = sender as PrototypePager;
+            switch (prevPageButtonVisibilityMode)
+            {
+                case ButtonVisibilityMode.Auto:
+                case ButtonVisibilityMode.AlwaysVisible:
+                    VisualStateManager.GoToState(pager, "PreviousPageButtonVisible", false);
+                    break;
+                case ButtonVisibilityMode.None:
+                    VisualStateManager.GoToState(pager, "PreviousPageButtonCollapsed", false);
+                    break;
+                case ButtonVisibilityMode.HiddenOnEdge:
+                    UpdateHiddenOnEdgeButtons(pager);
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        private static void OnFirstPageButtonVisibilityChanged(DependencyObject sender, DependencyPropertyChangedEventArgs args)
+        {
+            ButtonVisibilityMode firstPageButtonVisibilityMode = (ButtonVisibilityMode)args.NewValue;
+            PrototypePager pager = sender as PrototypePager;
+
+            switch (firstPageButtonVisibilityMode)
+            {
+                case ButtonVisibilityMode.Auto:
+                case ButtonVisibilityMode.AlwaysVisible:
+                    VisualStateManager.GoToState(pager, "FirstPageButtonVisible", false);
+                    break;
+                case ButtonVisibilityMode.None:
+                    VisualStateManager.GoToState(pager, "FirstPageButtonCollapsed", false);
+                    break;
+                case ButtonVisibilityMode.HiddenOnEdge:
+                    UpdateHiddenOnEdgeButtons(pager);
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        private void DisablePageButtonsOnEdge(DependencyObject sender, DependencyProperty dp)
+        {
+            FirstPageButton.IsEnabled = SelectedIndex != 1;
+            PreviousPageButton.IsEnabled = SelectedIndex != 1;
+            NextPageButton.IsEnabled = SelectedIndex != NumberOfPages;
+            LastPageButton.IsEnabled = SelectedIndex != NumberOfPages;
+            UpdateHiddenOnEdgeButtons(this);
+        }
+
+        private static void UpdateHiddenOnEdgeButtons(PrototypePager pager)
+        {
+            if (pager.FirstPageButtonVisibility == ButtonVisibilityMode.HiddenOnEdge)
+            {
+                if (pager.FirstPageButton.IsEnabled)
+                {
+                    VisualStateManager.GoToState(pager, "FirstPageButtonVisible", false);
+                }
+                else
+                {
+                    VisualStateManager.GoToState(pager, "FirstPageButtonCollapsed", false);
+                }
+            }
```

Comment:
I don't think this check is good enough.  What if the button is disabled for some reason other than being on the edge (like the last button is disabled because the page isn't ready yet). I think you need to check *ButtonVisibilityVisibility value and the selected index and use those to determine which visual state to be in.

Reply:
Hm, I did not think of this. I have added in a modification to handle this!